### PR TITLE
Task 5

### PR DESCRIPTION
**Pull request by SamoylenkVi**

### CODE REVIEW

File: project/src/components/AddReviewForm/index.tsx [hunk 1]
```diff
@@ -1,14 +1,34 @@
 import React, { useState } from 'react';
 import { RATING_FILM } from '../../constants';
+import { useAppDispatch } from '../../hooks';
+import { fetchAddNewComment } from '../../store/api-action';
 
-const ratings = Array.from({ length: RATING_FILM }, (_, index) => index + 1);
+const ratings = Array.from({ length: RATING_FILM }, (_, index) => index + 1).reverse();
```
Comment (Pavel-Liteiniy) [inline project/src/components/AddReviewForm/index.tsx]: Try to do the same without `.reverse()`
Tip: you have a length - `RATING_FILM`

File: project/src/components/AddReviewForm/index.tsx [hunk 2]
```diff
@@ -1,14 +1,34 @@
...
     commentText: '',
   });
 
+  const handleCreateComment = (evt: React.FormEvent<HTMLFormElement>) => {
+    evt.preventDefault();
+
+    dispatch(fetchAddNewComment({
+      comment: review.commentText,
+      rating: review.rating,
+      id: id,
```
Comment (Pavel-Liteiniy) [inline project/src/components/AddReviewForm/index.tsx]:
```suggestion
      id,
```

File: project/src/components/Authorization/components/logout.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { useAppDispatch } from '../../../hooks';
+import { fetchLogoutAction} from '../../../store/api-action';
+import '../style.scss';
+
+export const Logout = () => {
+  const dispatch = useAppDispatch();
+
+  const signOutHandler = (evt : React.MouseEvent) => {
```
Comment (Pavel-Liteiniy) [inline project/src/components/Authorization/components/logout.tsx]:
```suggestion
  const handleSignOutButtonClick = (evt : React.MouseEvent<HTMLButtonElement>) => {
```

Use consistent naming image

File: project/src/components/FilmDetailsReview/index.tsx [hunk 4]
```diff
@@ -1,15 +1,33 @@
...
+
+  useEffect(() => {
+    if(!id) {
+      return;
+    }
+    dispatch(fetchCommentsAction(id));
+
+  }, [id, dispatch]);
+
+  const comments = useAppSelector((state) => state.filmComments);
```
Comment (Pavel-Liteiniy) [inline project/src/components/FilmDetailsReview/index.tsx]: let's move `(state) => state.filmComments` to /selectors dir

File: project/src/components/FilmDetailsReview/index.tsx [hunk 5]
```diff
@@ -1,15 +1,33 @@
...
-  comments: Comment[];
+  id: number;
 }
 
-export const FilmDetailsReview = ({ comments }:FilmDetailsReviewProps) => {
+export const FilmDetailsReview = ({ id }:FilmDetailsReviewProps) => {
+  const dispatch = useAppDispatch();
+
+  useEffect(() => {
+    if(!id) {
```
Comment (Pavel-Liteiniy) [inline project/src/components/FilmDetailsReview/index.tsx]: Look's like this scenario is impossible.

```
type FilmDetailsReviewProps = {
  id: number;
}
```

File: project/src/components/FilmDetailsReview/index.tsx [hunk 6]
```diff
@@ -1,15 +1,33 @@
...
+  const comments = useAppSelector((state) => state.filmComments);
+
+  if(!comments) {
+    return null;
+  }
+
   const commentComponents = comments.map((comment) => (
     <FilmReview
-      comment={comment} key={comment.id}
+      commentData={comment} key={comment.id}
```
Comment (Pavel-Liteiniy) [inline project/src/components/FilmDetailsReview/index.tsx]:
```suggestion
      commentData={comment}
      key={comment.id}
```

File: project/src/components/FilmList/index.tsx [hunk 7]
```diff
@@ -4,15 +4,16 @@ import { FilmCard } from '../FilmCard/index';
 import { DEFAULT_GENRE, FILM_CARD_COUNT, QUERY_PARAM} from '../../constants';
 import { useAppSelector} from '../../hooks';
 import {ShowMoreButton} from '../ShowMoreButton';
+import { Loader } from '../Loader';
 import {Film} from '../../types/film';
 
-
 export const FilmList = () => {
   const films = useAppSelector((state) => state.films);
+  const isFilmsLoading = useAppSelector((state) => state.isFilmsDataLoading);
```
Comment (Pavel-Liteiniy) [inline project/src/components/FilmList/index.tsx]: Let's move to /selectors dir

File: project/src/components/FilmList/index.tsx [hunk 8]
```diff
@@ -4,15 +4,16 @@ import { FilmCard } from '../FilmCard/index';
...
 
   //TODO fix this;
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  const [activeCardId, setActiveCardId] = useState('');
+  const [activeCardId, setActiveCardId] = useState<number>();
```
Comment (Pavel-Liteiniy) [inline project/src/components/FilmList/index.tsx]: let's create type `CardId` in /types dir and use it here and in all other places

File: project/src/components/FilmList/index.tsx [hunk 9]
```diff
@@ -37,28 +38,30 @@ export const FilmList = () => {
...
-              name = {film.name}
-              previewImage = {film.previewImage}
-              previewVideoLink = {film.previewVideoLink}
-              id = {film.id}
-              onMouseEnter={handleMouseEnter}
-            />
-          ))
-        }
-      </div>
+    isFilmsLoading ? <Loader /> :
```
Comment (Pavel-Liteiniy) [inline project/src/components/FilmList/index.tsx]: Let's create early return

```
if (isFilmsLoading) {
  return <Loader />
}

return (
  <>
    <div className="catalog__films-list">
   ...
)
```

File: project/src/components/Logo/index.tsx [hunk 10]
```diff
@@ -1,9 +1,11 @@
+import {Link} from 'react-router-dom';
+
 export const Logo = () => (
   <div className="logo">
-    <a href="main.html" className="logo__link">
+    <Link to="/" className="logo__link">
```
Comment (Pavel-Liteiniy) [inline project/src/components/Logo/index.tsx]:
```suggestion
    <Link to={Route.Main} className="logo__link">
```

File: project/src/hooks/index.ts [hunk 11]
```diff
@@ -1,6 +1,32 @@
...
 export const useAppSelector: TypedUseSelectorHook<State> = useSelector;
+
+export const useScrollToTop = () => {
+  const {pathname} = useLocation();
+
+  useEffect(() => {
+    window.scrollTo({
+      top: 0,
+      left: 0,
+      behavior: 'smooth',
```
Comment (Pavel-Liteiniy) [inline project/src/hooks/index.ts]: why 'smooth' ?

File: project/src/pages/AddReviewPage/index.tsx [hunk 12]
```diff
@@ -29,14 +47,13 @@ export const AddReviewPage = () => {
 
         <header className="page-header">
           <Logo />
-
           <nav className="breadcrumbs">
             <ul className="breadcrumbs__list">
               <li className="breadcrumbs__item">
-                <Link to={'/films/2'} className="breadcrumbs__link">{name}</Link>
+                { id && <Link to={`/films/${id}/details`} className="breadcrumbs__link">{name}</Link> }
```
Comment (Pavel-Liteiniy) [inline project/src/pages/AddReviewPage/index.tsx]:
```suggestion
                { id && <Link to={generatePath(Route.FilmDetails, { id })} className="breadcrumbs__link">{name}</Link> }
```

https://reactrouter.com/en/main/utils/generate-path

File: project/src/pages/AddReviewPage/index.tsx [hunk 13]
```diff
@@ -29,14 +47,13 @@ export const AddReviewPage = () => {
...
               </li>
               <li className="breadcrumbs__item">
-                <a className="breadcrumbs__link">Add review</a>
+                <Link className="breadcrumbs__link" to={''}>Add review</Link>
```
Comment (Pavel-Liteiniy) [inline project/src/pages/AddReviewPage/index.tsx]: What `to={''}` means?

File: project/src/pages/AddReviewPage/index.tsx [hunk 14]
```diff
@@ -52,7 +69,10 @@ export const AddReviewPage = () => {
           <img src={previewImage} alt="The Grand Budapest Hotel poster" width="218" height="327" />
         </div>
       </div>
-      <AddReviewForm />
+      {
+        id && <AddReviewForm id={id}/>
```
Comment (Pavel-Liteiniy) [inline project/src/pages/AddReviewPage/index.tsx]: Check what happened if `id=0`

File: project/src/pages/FilmDetailsPage/index.tsx [hunk 15]
```diff
@@ -1,25 +1,37 @@
...
 import { Logo } from '../../components/Logo';
+import {fetchFilmAction, fetchSimilarFilmsAction} from '../../store/api-action';
+
 
 export const FilmDetailsPage = () => {
+  const dispatch = useAppDispatch();
   const { id } = useParams<string>();
-  const film = films.find((item) => item.id === id);
+
+  useEffect(() => {
```
Comment (Pavel-Liteiniy) [inline project/src/pages/FilmDetailsPage/index.tsx:19]: Let's fix unnesseccary requests  in all useEffect hooks
https://redux-toolkit.js.org/api/createAsyncThunk#canceling-while-running

File: project/src/pages/Login/index.tsx [hunk 16]
```diff
@@ -1,34 +1,74 @@
...
+                id="user-password"
+              />
+              <label className="sign-in__label visually-hidden" htmlFor="user-password">Password</label>
+            </div>
           </div>
-          <div className="sign-in__field">
-            <input className="sign-in__input" type="password" placeholder="Password" name="user-password" id="user-password" />
-            <label className="sign-in__label visually-hidden" htmlFor="user-password">Password</label>
+          <div className="sign-in__submit">
+            <button className="sign-in__btn" type="submit">Sign in</button>
```
Comment (Pavel-Liteiniy) [inline project/src/pages/Login/index.tsx]: let's disable button while fetchLoginAction in progress

File: project/src/store/api-action.ts [hunk 17]
```diff
@@ -19,3 +26,97 @@ export const fetchFilmsAction = createAsyncThunk<
...
+  extra: AxiosInstance;
+}>(
+  'data/fetchSimilarFilm',
+  async (id, { extra: api }) => {
+    const { data } = await api.get<Film[]>(`${API_ROUTE.Films}/${id}/similar`);
+    return data;
+  },
+);
+
+export const fetchCommentsAction = createAsyncThunk<
```
Comment (Pavel-Liteiniy) [inline project/src/store/api-action.ts]:
```suggestion
export const fetchComments = createAsyncThunk<
```

let's avoid using "Action" postfix everywhere

File: project/src/store/store.ts [hunk 18]
```diff
@@ -1,14 +1,16 @@
...
+
 export const store = configureStore({
   reducer: reducer,
   middleware: (getDefaultMiddleware) =>
     getDefaultMiddleware({
       thunk: {
         extraArgument: api,
       },
-    }),
+    }).concat(redirectM),
```
Comment (Pavel-Liteiniy) [inline project/src/store/store.ts]: what's M in redirect**M** mean?

File: project/src/components/FilmList/index.tsx (hunk 8 above)
Comment (Pavel-Liteiniy) [reply project/src/components/FilmList/index.tsx]: Or if we have some "Card" type let's use it like `useState<Card['id']>();`

**Review by Pavel-Liteiniy — CHANGES_REQUESTED** [review]

File: project/src/components/Authorization/components/logout.tsx (hunk 3 above)
Comment (SamoylenkVi) [reply project/src/components/Authorization/components/logout.tsx]: fixed

File: project/src/store/api-action.ts (hunk 17 above)
Comment (SamoylenkVi) [reply project/src/store/api-action.ts]: fixed

File: project/src/store/store.ts (hunk 18 above)
Comment (SamoylenkVi) [reply project/src/store/store.ts]: fixed

File: project/src/components/AddReviewForm/index.tsx (hunk 2 above)
Comment (SamoylenkVi) [reply project/src/components/AddReviewForm/index.tsx]: fixed

File: project/src/components/FilmDetailsReview/index.tsx (hunk 5 above)
Comment (SamoylenkVi) [reply project/src/components/FilmDetailsReview/index.tsx]: fixed

File: project/src/pages/AddReviewPage/index.tsx (hunk 14 above)
Comment (SamoylenkVi) [reply project/src/pages/AddReviewPage/index.tsx]: fixid

File: project/src/components/FilmDetailsReview/index.tsx (hunk 4 above)
Comment (SamoylenkVi) [reply project/src/components/FilmDetailsReview/index.tsx]: fixed

File: project/src/components/FilmList/index.tsx (hunk 7 above)
Comment (SamoylenkVi) [reply project/src/components/FilmList/index.tsx]: fixed

File: project/src/components/FilmDetailsReview/index.tsx (hunk 6 above)
Comment (SamoylenkVi) [reply project/src/components/FilmDetailsReview/index.tsx]: fixed

File: project/src/components/FilmList/index.tsx (hunk 8 above)
Comment (SamoylenkVi) [reply project/src/components/FilmList/index.tsx]: fixed

File: project/src/components/FilmList/index.tsx (hunk 9 above)
Comment (SamoylenkVi) [reply project/src/components/FilmList/index.tsx]: fixed

File: project/src/components/Logo/index.tsx (hunk 10 above)
Comment (SamoylenkVi) [reply project/src/components/Logo/index.tsx]: fixed

File: project/src/hooks/index.ts (hunk 11 above)
Comment (SamoylenkVi) [reply project/src/hooks/index.ts]: fixed

File: project/src/pages/AddReviewPage/index.tsx (hunk 13 above)
Comment (SamoylenkVi) [reply project/src/pages/AddReviewPage/index.tsx]: fixed

File: project/src/pages/Login/index.tsx (hunk 16 above)
Comment (SamoylenkVi) [reply project/src/pages/Login/index.tsx]: fixed

File: project/src/components/AddReviewForm/index.tsx (hunk 1 above)
Comment (SamoylenkVi) [reply project/src/components/AddReviewForm/index.tsx]: fixed

File: project/src/pages/Login/index.tsx (hunk 16 above)
Comment (SamoylenkVi) [reply project/src/pages/Login/index.tsx]: fixed

File: project/src/pages/FilmDetailsPage/index.tsx (hunk 15 above)
Comment (SamoylenkVi) [reply project/src/pages/FilmDetailsPage/index.tsx:19]: fixed

File: project/src/components/AddReviewForm/index.tsx (hunk 1 above)
Comment (SamoylenkVi) [reply project/src/components/AddReviewForm/index.tsx]: fixed

File: project/src/pages/Login/index.tsx (hunk 16 above)
Comment (SamoylenkVi) [reply project/src/pages/Login/index.tsx]: fixed

File: project/src/pages/AddReviewPage/index.tsx (hunk 12 above)
Comment (SamoylenkVi) [reply project/src/pages/AddReviewPage/index.tsx]: fixed

File: project/src/pages/Login/index.tsx [hunk 19]
```diff
@@ -16,7 +16,7 @@ export const Login = () => {
     dispatch(fetchLoginAction(loginData));
   };
 
-  const submitLoginFormHandle = (evt: React.FormEvent<HTMLFormElement>) => {
+  const handleSubmitLoginForm = (evt: React.FormEvent<HTMLFormElement>) => {
```
Comment (Pavel-Liteiniy) [inline project/src/pages/Login/index.tsx]: handle + element + action
check others please 🙃

```
handleLoginFormSubmit
```

File: project/src/components/ShowMoreButton/index.tsx [hunk 20]
```diff
@@ -4,13 +4,13 @@ type ShowMoreButtonProps = {
 
 export const ShowMoreButton = ({onClick}:ShowMoreButtonProps) => {
 
-  const clickButtonHandler = () => {
+  const handleClickButton = () => {
```
Comment (Pavel-Liteiniy) [inline project/src/components/ShowMoreButton/index.tsx]: handleButtonClick

**Review by Pavel-Liteiniy — CHANGES_REQUESTED** [review]

**Comment by Pavel-Liteiniy** [comment]
Approve! Good job 🤘

File: project/src/pages/Login/index.tsx (hunk 19 above)
Comment (SamoylenkVi) [reply project/src/pages/Login/index.tsx]: fixed

File: project/src/components/ShowMoreButton/index.tsx (hunk 20 above)
Comment (SamoylenkVi) [reply project/src/components/ShowMoreButton/index.tsx]: fixed

**Pull request merged**